models/turbine_models/custom_models/sd_inference/sd_pipeline.py

```diff
@@ -0,0 +1,748 @@
+        scheduler_id: str = "EulerDiscrete",
+        return_imgs: bool = False,
+    ):
+        needs_new_scheduler = (
+            (steps and steps != self.num_inference_steps)
+            or (cpu_scheduling != self.cpu_scheduling)
+            and self.split_scheduler
+        )
+        if not self.scheduler and not self.compiled_pipeline:
+            needs_new_scheduler = True
+
+        if guidance_scale == 0:
+            negative_prompt = prompt
+            prompt = ""
+
+        self.cpu_scheduling = cpu_scheduling
+        if steps and needs_new_scheduler:
+            self.num_inference_steps = steps
+            self.load_scheduler(scheduler_id, steps)
+
+        pipe_start = time.time()
+        numpy_images = []
+
+        samples = self.get_rand_latents(seed, batch_count)
+
+        # Tokenize prompt and negative prompt.
+        if self.is_sdxl:
+            prompt_embeds, negative_embeds = self.encode_prompts_sdxl(
+                prompt, negative_prompt
+            )
+        else:
+            prompt_embeds, negative_embeds = encode_prompt(
+                self, prompt, negative_prompt
+            )
+
+        for i in range(batch_count):
+            produce_latents_input = [
+                samples[i],
+                prompt_embeds,
+                negative_embeds,
+                steps,
+                guidance_scale,
+            ]
+            if self.is_sdxl:
+                latents = self._produce_latents_sdxl(*produce_latents_input)
+            else:
+                latents = self._produce_latents_sd(*produce_latents_input)
+            image = self.vae("decode", [latents])
+            numpy_images.append(image)
+            pipe_end = time.time()
+
+        logging.info(f"Total inference time: {pipe_end - pipe_start:.2f}s")
+        timestamp = dt.now().strftime("%Y-%m-%d_%H-%M-%S")
+        images = []
+        for idx, image in enumerate(numpy_images):
+            image = torch.from_numpy(image).cpu().permute(0, 2, 3, 1).float().numpy()
+            image = numpy_to_pil_image(image)
+            images.append(image[0])
+        if return_imgs:
+            return images
+        for idx, image in enumerate(images):
+            img_path = "sdxl_output_" + timestamp + "_" + str(idx) + ".png"
+            image.save(img_path)
+            print(img_path, "saved")
+        return
+
+
+def numpy_to_pil_image(images):
+    """
+    Convert a numpy image or a batch of images to a PIL image.
+    """
+    if images.ndim == 3:
+        images = images[None, ...]
+    images = (images * 255).round().astype("uint8")
+    if images.shape[-1] == 1:
+        # special case for grayscale (single channel) images
+        pil_images = [Image.fromarray(image.squeeze(), mode="L") for image in images]
+    else:
+        pil_images = [Image.fromarray(image) for image in images]
+
+    return pil_images
+
+
+if __name__ == "__main__":
+    from turbine_models.custom_models.sd_inference.sd_cmd_opts import args
```

Comment:
Should we use sd_cmd_opts here? The default height and width is 512.

Reply:
I guess it's ok, can just pass in height and width as 1024 through the CL args for SDXL.

Reply:
I kept this as a default since it is the only size supported by all the pipeline's supported models. Since the pipelines are using the same class/API now, they should either use the same args or we need a layer between command line arguments and pipeline API. If we want to do the latter it would be best as a follow-up.

I plan on doing a scrub of things like print statements and commented code as a follow-up -- there are also a few files/functions to deprecate or remove, so that can come as a follow-up to this patch.